physical/raft/snapshot.go

```diff
@@ -151,69 +156,154 @@ func (f *BoltSnapshotStore) getBoltSnapshotMeta() (*raft.SnapshotMeta, error) {
+	// Stream data out of the FSM to calculate the size
+	readCloser, writeCloser := io.Pipe()
+	metaReadCloser, metaWriteCloser := io.Pipe()
+	go func() {
+		f.fsm.writeTo(context.Background(), metaWriteCloser, writeCloser)
+	}()
+
+	// Compute the size
+	n, err := io.Copy(ioutil.Discard, metaReadCloser)
+	if err != nil {
+		f.logger.Error("failed to read state file", "error", err)
+		metaReadCloser.Close()
+		readCloser.Close()
+		return nil, nil, err
+	}
+
+	meta.Size = n
+
+	return meta, readCloser, nil
+}
+
+func (f *BoltSnapshotStore) getMetaFromDB(id string) (*raft.SnapshotMeta, error) {
+	if len(id) == 0 {
+		return nil, errors.New("can not open empty snapshot ID")
+	}
+
+	filename := filepath.Join(f.path, id, databaseFilename)
+	boltDB, err := bolt.Open(filename, 0666, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, err
+	}
+	defer boltDB.Close()
+
+	meta := &raft.SnapshotMeta{
+		Version: 1,
+		ID:      id,
+	}
+
+	err = boltDB.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(configBucketName)
+		val := b.Get(latestIndexKey)
+		if val != nil {
+			var snapshotIndexes IndexValue
+			err := proto.Unmarshal(val, &snapshotIndexes)
+			if err != nil {
+				return err
+			}
+
+			meta.Index = snapshotIndexes.Index
+			meta.Term = snapshotIndexes.Term
 		}
 
-		readCloser = &boltSnapshotMetadataReader{
-			meta:       meta,
-			ReadCloser: readCloser,
+		// Read in our latest config and populate it inmemory
+		val = b.Get(latestConfigKey)
+		if val != nil {
+			var config ConfigurationValue
+			err := proto.Unmarshal(val, &config)
+			if err != nil {
+				return err
+			}
+
+			meta.ConfigurationIndex, meta.Configuration = protoConfigurationToRaftConfiguration(&config)
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return meta, nil
+}
+
+func (f *BoltSnapshotStore) openFromFile(id string) (*raft.SnapshotMeta, io.ReadCloser, error) {
+	// TODO: should we insted have a separate metadata file so we can checksum
+	// the db file?
+	meta, err := f.getMetaFromDB(id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	filename := filepath.Join(f.path, id, databaseFilename)
+	readCloser := &boltSnapshotInstaller{
+		meta:       meta,
+		ReadCloser: ioutil.NopCloser(strings.NewReader(filename)),
+		filename:   filename,
 	}
 
 	return meta, readCloser, nil
 }
 
 // ReapSnapshots reaps any snapshots beyond the retain count.
 func (f *BoltSnapshotStore) ReapSnapshots() error {
-	return f.fileSnapStore.ReapSnapshots()
+	snapshots, err := ioutil.ReadDir(f.path)
+	switch {
+	case err == nil:
+	case os.IsNotExist(err):
+		return nil
+	default:
+		f.logger.Error("failed to scan snapshot directory", "error", err)
+		return err
+	}
+
+	for _, snap := range snapshots {
+		// Ignore any files
+		if !snap.IsDir() {
+			continue
+		}
+
+		// Ignore any temporary snapshots
+		dirName := snap.Name()
+		if strings.HasSuffix(dirName, tmpSuffix) {
+			f.logger.Warn("found temporary snapshot", "name", dirName)
```

Comment:
Do you need a `continue` here if you're wanting to ignore temporary snapshots?

Reply:
Good question, we want to clean them up as well. Looks like the comment is wrong. Will update.